### ISSUE TITLE
Ergonomics of dealing with and fixing up various failures

### ISSUE THREAD

**Issue by kamrann** [build2/build2#426]
One recurring pain point I have with `build2` relates to recovery when something goes wrong, be it because of user error, invalid project setup, or a toolchain bug. In particular there are a few aspects which I'm wondering if there's potential for improvement (and if there may be some solution or workaround of which I'm just not aware).

### 1. Errors emitted after a potentially long fetch cycle.
I have rather a lot of small packages hosted on `git`, as well as multiple projects and configurations, so I suspect my fetch times are much longer than most and perhaps others don't have this trouble. Even without a fetch, just disfigure-reconfigure can be a non-trivial wait. It's common for some changes I make to trigger a fetch and/or sync which can take quite a while, and only after that process is done am I presented with an error. This is particularly painful for `bdep init` since it needs to do the full fetch again once I've addressed the issue. 

Obviously in some cases this is inevitable, but sometimes the nature of the error makes me suspect it could have potentially been detected before doing the sync. I realize this is a bit vague, I'll try to pay closer attention next time I notice it and give some more specific example.

### 2. Robustness of packages w.r.t. dependencies.
Say I may make a bad change in a `buildfile` and trigger a sync. An error will be emitted, but it seems to often be the case that after emitting the error, it will then report `info: package something-downstream is now broken; use 'pkg-purge' to remove`. This seems somewhat heavy, but I don't know how intentional it is. Are there fundamental reasons why sometimes such packages can't be just left as-is until the upstream error is addressed, rather than being considered 'broken'?

### 3. Recovering from broken package state. 
Following the prompts as a beginner/someone with a bad memory, we'd go through the following process:
```
info: package some-pkg is now broken; use 'pkg-purge' to remove

> cd into/build/config
> bpkg pkg-purge some-pkg
error: broken package some-pkg can only be purged with --force
> bpkg pkg-purge --force some-pkg
error: output directory of broken package some-pkg still exists
  info: remove .\some-pkg-...\ manually then re-run pkg-purge
> rm ...
> bpkg pkg-purge --force some-pkg
purged some-pkg/...
```
This just feels a bit cumbersome. But perhaps the issue is really the frequency with which I run into (2) and similar. If broken packages were rare then this wouldn't really be a concern. Though I guess at least specifying the requirement of `--force` in the initial diagnostic would be worthwhile.

**Comment by kamrann** [comment]
Regarding (1), quickly ran into an example that is fairly common for me:
```
error: initialized package some-pkg specified as dependency
```
Now, in practice think this error for me pretty much always gets triggered as a result of what is a known bug (#399); that said, this seems like an example of something that should probably be detectable up front and wouldn't be affected by the results of the fetch?

**Comment by boris-kolpackov** [comment]
> but sometimes the nature of the error makes me suspect it could have potentially been detected before doing the sync

Could be in certain cases. But in others we will likely instead have a false positive, or detect something else unrelated, confusing the user even more.

My feeling is that this machinery has gotten so complex that trying to introduce some heuristics that sometimes try to guess something based on some transitory state will only make matters worse.

I think the solution for this specific issue is better fetch caching (so that on the subsequent run the fetch part is fast): https://github.com/build2/build2/issues/378

> Are there fundamental reasons why sometimes such packages can't be just left as-is until the upstream error is addressed, rather than being considered 'broken'?

Yes, we have failed to configure the package and it's left in the unconfigured state. Again, I think attempting to continue using a package in such a state will only lead to further confusion.

Regarding recovering from the broken state, we have the long-standing plan to have something like `bdep-reset`:

https://github.com/build2/build2/issues/106
https://github.com/build2/build2/issues/257

Overall, I agree that there are quite a few sharp edges that we would like to try and smooth out.

**Comment by kamrann** [comment]
> My feeling is that this machinery has gotten so complex that trying to introduce some heuristics that sometimes try to guess something based on some transitory state will only make matters worse.
> 
> I think the solution for this specific issue is better fetch caching (so that on the subsequent run the fetch part is fast): #378

Understood. And yes, for sure if fetch were faster than such things would largely become a non-issue.

> Yes, we have failed to configure the package and it's left in the unconfigured state. Again, I think attempting to continue using a package in such a state will only lead to further confusion.

Okay. Something about the behaviour here was still nagging at me when reading this, but I was struggling to pinpoint it. I'll have a go though. 

To give an example that I've hit a few times, occasionally I've triggered a sync forgetting that I have an executable from the proejct still running. Now this may be a Windows thing, but it's very fussy about file access to dlls/pdbs, so the reconfigure can fail with `error: unable to remove file ...`, followed by the statement about the package being broken. Investigating the state of the configuration after this seems to show that the package that failed is left `broken`, whilst any downstream dependents are in an `unpacked` state. Now as mentioned the process for purging and reinitializing the broken package is a bit longwinded, but after having done so I can apparently reconfigure everything without issue, and without having to do anything similar for those downstream unpacked packages. Which just makes me wonder, what is the benefit of the `broken` state, as opposed to just having left that package in an `unpacked` state along with all its dependents, which is seemingly a much simpler situation to recover from?

Anyway, since no-one else has weighed in or seems to be suffering too much from these things, feel free to close this if you think everything of relevance is already covered by the other issues.

**Comment by boris-kolpackov** [comment]
> And yes, for sure if fetch were faster than such things would largely become a non-issue.

FYI, fetch cache is available in the latest staged toolchain (and will be available in the `0.18.0` release). For details, see:

https://build2.org/release/0.18.0.xhtml#bpkg-fetch-cache
https://build2.org/release/0.18.0.xhtml#bdep-fetch-cache-offline
